src/Sarif/Visitors/PartitioningVisitor.cs

```diff
@@ -0,0 +1,454 @@
+                    if (!artifactIndices.Contains(node.Index))
+                    {
+                        artifactIndices.Add(node.Index);
+
+                        Artifact artifact = currentRunArtifacts[node.Index];
+                        while (artifact.ParentIndex >= 0)
+                        {
+                            artifactIndices.Add(artifact.ParentIndex);
+                            artifact = currentRunArtifacts[artifact.ParentIndex];
+                        }
+                    }
+                }
+            }
+
+            return base.VisitArtifactLocation(node);
+        }
+
+        private void CreatePartitionLogDictionary()
+        {
+            partitionLogDictionary = partitionValues.ToDictionary(
+                keySelector: pv => pv,
+                elementSelector: CreatePartitionLog);
+        }
+
+        private SarifLog CreatePartitionLog(T partitionValue)
+        {
+            SarifLog partitionLog;
+            if (deepClone)
+            {
+                // Save time and space by not cloning the runs unless and until necessary. We will
+                // only need to clone the runs that have results in this partition.
+                IList<Run> originalRuns = originalLog.Runs;
+                originalLog.Runs = null;
+
+                partitionLog = originalLog.DeepClone();
+
+                originalLog.Runs = originalRuns;
+            }
+            else
+            {
+                partitionLog = new SarifLog(originalLog)
+                {
+                    Runs = null
+                };
+            }
+
+            partitionLog.Runs = new List<Run>();
+            partitionLog.SetProperty(PartitionValuePropertyName, partitionValue);
+
+            var artifactIndexRemappingDictionaries = new List<Dictionary<int, int>>();
+            for (int iOriginalRun = 0; iOriginalRun < partitionResultDictionaries.Count; ++iOriginalRun)
+            {
+                // Are there results for this run in this partition?
+                if (partitionResultDictionaries[iOriginalRun].TryGetValue(partitionValue, out List<Result> results))
+                {
+                    // Yes, so we'll need a copy of the original run in which the results, and
+                    // certain run-level collections such as Run.Artifacts, have been replaced.
+                    Run originalRun = originalLog.Runs[iOriginalRun];
+                    Run partitionRun;
+                    if (deepClone)
+                    {
+                        // Save time and space by only cloning the necessary results and associated
+                        // collection elements. We already cloned the relevant results in VisitResult.
+                        IList <Result> originalResults = originalRun.Results;
+                        IList<Artifact> originalArtifacts = originalRun.Artifacts;
+                        originalRun.Results = null;
```

Comment:
I think a function wrapping the "DeepCloneExceptXAndY" functionality would help. It took me a moment to figure out that's what you're doing (storing, clearing, cloning, and then resetting the references you don't want cloned)

Reply:
It's a nice idea but the function name (which would now be `DeepCloneExceptResultsAndArtifacts`) will become unwieldy once it's `DeepCloneExceptResultArtifactsGraphsLogicalLocationsThreadFlowLocationsWebRequestsAndWebResponses` :-)

---
In reply to: [357431653](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357431653) [](ancestors = 357431653)

Reply:
Oh, I suppose `DeepCloneExceptRunLevelCollections` would be ok. Let me think about it.

---
In reply to: [357441324](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357441324) [](ancestors = 357441324,357431653)

Reply:
No, the comment already says we're "cloning only the necessary results and associated collection elements." I'm going to leave this.

---
In reply to: [357442600](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357442600) [](ancestors = 357442600,357441324,357431653)